core/metacling/src/TCling.cxx

```diff
@@ -3574,10 +3574,14 @@ void TCling::SetClassInfo(TClass* cl, Bool_t reload)
    cl->fClassInfo = 0;
    std::string name(cl->GetName());
 
+   bool useCxxModules = false;
+#ifdef R__USE_CXXMODULES
+   useCxxModules = true;
+#endif
    // Handle the special case of 'tuple' where we ignore the real implementation
    // details and just overlay a 'simpler'/'simplistic' version that is easy
    // for the I/O to understand and handle.
-   if (strncmp(cl->GetName(),"tuple<",strlen("tuple<"))==0) {
+   if (!useCxxModules && strncmp(cl->GetName(),"tuple<",strlen("tuple<"))==0) {
```

Comment:
Why do we enter this routine N times causing the redefinition errors?

Reply:
This is called only once, but it looks that Transactions are divided into pieces because of RAIIs and InternalParser where it creates Virtual FileID is called several times because of that.

Reply:
Are you talking about he cache in the LookupHelper?

Reply:
the current algorithm is as follow:
```
if (io requested for std::ntuple< param_set >) then
    Declare("#ifndef ROOT_INTERNAL_TEmulated_mangled_name_of_emulated_tuple
          #define ROOT_INTERNAL_TEmulated_mangled_name_of_emulated_tuple
          struct ROOT::TEmulatedTuple< param_set > { ... }
          #endif");
endif
```
This needs to be done at run-time and can be done at rootcling time to prevent auto-parsing (and store the TProtoClass info about the TEmulatedTuple in the rootpcm).

If 2 different dictionary uses for I/O the same std::ntuple, the TEmulatedTuple will end up in both module file.

For **some** reason, when those are stored in the clang pcm then do **not** respect the #ifndef and thus lead to a redeclaration.

Since with the clang-pcm, auto-parsing will be cheaper, we could get away with delaying the instantiation at run-time (But still it is likely to be more expansive in cpu and memory).

Cheers,
Philippe.